core/src/main/java/org/apache/druid/data/input/IntermediateRowParsingReader.java

```diff
@@ -96,23 +99,35 @@ public void close() throws IOException
   public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
   {
     return intermediateRowIterator().map(row -> {
-      final Map<String, Object> rawColumns;
+
+      final List<Map<String, Object>> rawColumnsList;
       try {
-        rawColumns = toMap(row);
+        rawColumnsList = toMap(row);
       }
       catch (Exception e) {
-        return InputRowListPlusRawValues.of(null, new ParseException(e, "Unable to parse row [%s] into JSON", row));
+        return Collections.singletonList(InputRowListPlusRawValues.of(null,
+                                                                      new ParseException(e, "Unable to parse row [%s] into JSON", row)));
       }
+
+      List<InputRow> rows;
       try {
-        return InputRowListPlusRawValues.of(parseInputRows(row), rawColumns);
+        rows = parseInputRows(row);
       }
       catch (ParseException e) {
-        return InputRowListPlusRawValues.of(rawColumns, e);
+        return Collections.singletonList(InputRowListPlusRawValues.of(rawColumnsList.isEmpty() ? null : rawColumnsList.get(0), e));
```

Comment:
Should we return the entire `rawColumnsList` instead of the first element? I thought `InputRowListPlusRawValues` will be able to have a list of `rawColumns`.

Reply:
Also, should we fail if `rawColumnsList.isEmpty()`?

Reply:
To return the entire `rawColumnsList` here involves changes on `InputRowListPlusRawValues` which are heavily used by test cases. 

I checked the code again and find that there's no need to return raw column object when exception occurs. Because `InputSourceSample` re-throws exception once `InputRowListPlusRawValues.getParseException` returns exception.  

So the new commit passes a `null` to `InputRowListPlusRawValues` to eliminate this strange code here.

Reply:
> I checked the code again and find that there's no need to return raw column object when exception occurs. Because `InputSourceSample` re-throws exception once `InputRowListPlusRawValues.getParseException` returns exception.

[Actually, there is a clause catching `ParseException`](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/InputSourceSampler.java#L142-L146) in the sampler (this is a bad pattern by the way. We should refactor it at some point). In this case, the `rawColumns` is returned to show the raw data in the UI. I think the entire `rawColumnsList` should be returned to keep the current behaviour.

Reply:
There are some test failing in `InputSourceSamplerTest`. It seems like because of this change.

Reply:
It's been fixed. Now it's ready for final review.